vllm/model_executor/sampling_metadata.py

```diff
@@ -134,13 +134,62 @@ def __init__(
         num_prompts: int,
         skip_sampler_cpu_output: bool = False,
         reuse_sampling_tensors: bool = False,
+        # Used when multi-step is enabled with chunked-prefill. Refer to
+        # the comment in prepare_multistep_tensors.
+        selected_token_indices_multistep: Optional[torch.Tensor] = None
     ) -> None:
         self.seq_groups = seq_groups
         self.selected_token_indices = selected_token_indices
         self.categorized_sample_indices = categorized_sample_indices
         self.num_prompts = num_prompts
         self.skip_sampler_cpu_output = skip_sampler_cpu_output
         self.reuse_sampling_tensors = reuse_sampling_tensors
+        self.selected_token_indices_multistep = selected_token_indices_multistep
+
+    def prepare_multistep_tensors(self, num_queries: int, device: str,
+                                  pin_memory: bool):
+        """
+        Invoked when Multi-Step is enabled with Chunked-Prefill. 
+        When Multi-Step is enabled with Chunked-Prefill, the prompts and
+        decodes are scheduled together.
+        self.selected_token_indices is constructed for the first-step in
+        Multi-Step. However, the scheduled prompts, are fully processed
+        in the first-step and are processed as decodes in the rest of the steps.
+        This function prepares a "selected_token_indices" to be used
+        in the rest of the steps.
+
+        Example:
+        Let 2 prompts and 2 decodes be scheduled together. Let the
+        num-tokens to process for the 2 prompts be 5 and 8 resply.
+
+        In that case, self.sampled_token_indices will be,
+        [5, 13, 14, 15] as it is constructed for the first-step in
+        multi-step.
+        However, the prompts turns to decodes after the first-step
+        and the num-tokens for the previously-prompt sequences will
+        be 1 and 1 as they are decodes now. The self.sampled_token_indices
+        must be updated to [1,2,3,4].
+        prepare_multistep_tensors prepares the "selected_token_indices"
+        to be used in steps 2-N.
```

Comment:
TODO (varun) : The arrays are wrong. should be [4, 12, 13, 14] and [0, 1, 2, 3] resply.